worker/src/synchronizer.rs

```diff
@@ -133,55 +136,80 @@ impl Synchronizer {
-                            .duration_since(UNIX_EPOCH)
-                            .expect("Failed to measure time")
-                            .as_millis();
+                        let mut missing = HashSet::new();
+                        let mut available = HashSet::new();
 
-                        let mut missing = Vec::new();
-                        for digest in digests {
+                        for digest in digests.iter() {
                             // Ensure we do not send twice the same sync request.
-                            if self.pending.contains_key(&digest) {
+                            if self.pending.contains_key(digest) {
                                 continue;
                             }
 
                             // Check if we received the batch in the meantime.
-                            match self.store.read(digest).await {
+                            match self.store.read(*digest).await {
                                 Ok(None) => {
-                                    missing.push(digest);
+                                    missing.insert(*digest);
                                     debug!("Requesting sync for batch {digest}");
                                 },
                                 Ok(Some(_)) => {
                                     // The batch arrived in the meantime: no need to request it.
+                                    available.insert(*digest);
+                                    debug!("Digest {digest} already in store, nothing to sync");
+                                    continue;
                                 },
                                 Err(e) => {
                                     error!("{e}");
                                     continue;
                                 }
-                            }
+                            };
+                        }
 
-                            // Add the digest to the waiter.
-                            let deliver = digest;
-                            let (tx_cancel, rx_cancel) = mpsc::channel(1);
-                            let fut = Self::waiter(digest, self.store.clone(), deliver, rx_cancel);
-                            waiting.push(fut);
-                            self.pending.insert(digest, (self.round, tx_cancel, now));
+                        // reply back immediately for the available ones
+                        if !available.is_empty() {
+                            // Doing this will ensure the batch id will be populated to primary even
+                            // when other processes fail to do so (ex we received a batch from a peer
+                            // worker and message has been missed by primary).
+                            for digest in available {
+                                let message = WorkerPrimaryMessage::OthersBatch(digest, self.id);
+                                _ = self.tx_primary.send(message).await.tap(|err|{
```

Comment:
The `send` returns a `Result<(), SendError<T>>` so I think you want to use `tap_err` here, as this closure would be applied on the whole `Result` otherwise.

Reply:
Good point!